protoc-gen-swagger/defs.bzl

```diff
@@ -0,0 +1,79 @@
+def _collect_includes(srcs):
+    includes = ["."]
+    for src in srcs:
+        include = src.dirname
+        if include and not include in includes:
+            includes += [include]
+
+    return includes
+
+def _run_proto_gen_swagger(proto, transitive_proto_srcs, actions, protoc, protoc_gen_swagger):
+    swagger_file = actions.declare_file(
+        "%s.swagger.json" % proto.basename[:-len(".proto")],
+        sibling = proto,
+    )
+
+    args = actions.args()
+    args.add("--plugin=%s" % protoc_gen_swagger.path)
+    args.add("--swagger_out=logtostderr=true:%s" % swagger_file.dirname)
+    args.add("-Iexternal/com_google_protobuf/src")
+    args.add("-Iexternal/com_github_googleapis_googleapis")
+    args.add(["-I%s" % include for include in _collect_includes(transitive_proto_srcs)])
+    args.add(proto.basename)
+
+    actions.run(
+        executable = protoc,
+        inputs = transitive_proto_srcs + [protoc_gen_swagger],
+        outputs = [swagger_file],
+        arguments = [args],
+    )
+
+    return swagger_file
+
+def _proto_gen_swagger_impl(ctx):
+    transitive_proto_srcs = depset([ctx.file.proto, ctx.file._annotations] + ctx.files._well_known_protos)
+    for dep in ctx.attr.deps:
+        transitive_proto_srcs = depset(transitive=[transitive_proto_srcs, dep.proto.transitive_sources])
+
+    return struct(
+        files=depset([
+            _run_proto_gen_swagger(
+                ctx.file.proto,
+                transitive_proto_srcs.to_list(),
+                ctx.actions,
+                ctx.executable._protoc,
+                ctx.executable._protoc_gen_swagger,
+            )
+        ])
+    )
+
+protoc_gen_swagger = rule(
+    attrs = {
+        "proto": attr.label(
+            mandatory = True,
+            allow_single_file = True,
```

Comment:
You will need to remove this to accept a Label instead of files. From that Label you will extract the `src` files using a Provider.

Reply:
Donezo